pkg/kubelet/volume/iscsi/iscsi.go

```diff
@@ -0,0 +1,157 @@
+}
+
+func (plugin *ISCSIDiskPlugin) CanSupport(spec *api.Volume) bool {
+	if spec.ISCSIDisk != nil {
+		return true
+	}
+	exe := exec.New()
+	// see if iscsiadm is there
+	command := exe.Command("iscsiadm", []string{"-h"}...)
+	_, err := command.CombinedOutput()
+	if err == nil {
+		return true
+	}
+	return false
+}
+
+func (plugin *ISCSIDiskPlugin) NewBuilder(spec *api.Volume, podRef *api.ObjectReference) (volume.Builder, error) {
+	// Inject real implementations here, test through the internal function.
+	return plugin.newBuilderInternal(spec, podRef.UID, &ISCSIDiskUtil{}, mount.New(), exec.New())
+}
+
+func (plugin *ISCSIDiskPlugin) newBuilderInternal(spec *api.Volume, podUID types.UID, manager DiskManager, mounter mount.Interface, exe exec.Interface) (volume.Builder, error) {
+	lun := "0"
+	if spec.ISCSIDisk.Lun != 0 {
+		lun = strconv.Itoa(spec.ISCSIDisk.Lun)
+	}
+
+	return &iscsiDisk{
+		podUID:   podUID,
+		volName:  spec.Name,
+		portal:   spec.ISCSIDisk.TargetPortal,
+		iqn:      spec.ISCSIDisk.IQN,
+		lun:      lun,
+		fsType:   spec.ISCSIDisk.FSType,
+		readOnly: spec.ISCSIDisk.ReadOnly,
+		exec:     exe,
+		manager:  manager,
+		mounter:  mounter,
+		plugin:   plugin,
+	}, nil
+}
+
+func (plugin *ISCSIDiskPlugin) NewCleaner(volName string, podUID types.UID) (volume.Cleaner, error) {
+	// Inject real implementations here, test through the internal function.
+	return plugin.newCleanerInternal(volName, podUID, &ISCSIDiskUtil{}, mount.New(), exec.New())
+}
+
+func (plugin *ISCSIDiskPlugin) newCleanerInternal(volName string, podUID types.UID, manager DiskManager, mounter mount.Interface, exe exec.Interface) (volume.Cleaner, error) {
+
+	return &iscsiDisk{
+		podUID:  podUID,
+		volName: volName,
+		manager: manager,
+		mounter: mounter,
+		plugin:  plugin,
+		exec:    exe,
+	}, nil
+}
+
+type iscsiDisk struct {
+	volName  string
+	podUID   types.UID
+	portal   string
+	iqn      string
+	readOnly bool
+	lun      string
+	fsType   string
+	plugin   *ISCSIDiskPlugin
+	mounter  mount.Interface
+	// Utility interface that provides API calls to the provider to attach/detach disks.
+	manager DiskManager
+	exec    exec.Interface
+}
+
+func (iscsi *iscsiDisk) GetPath() string {
+	name := ISCSIDiskPluginName
+	// safe to use PodVolumeDir now: volume teardown occurs before pod is cleaned up
+	return iscsi.plugin.host.GetPodVolumeDir(iscsi.podUID, volume.EscapePluginName(name), iscsi.volName)
+}
+
+func (iscsi *iscsiDisk) SetUp() error {
+	err := CommonPDSetUp(iscsi.manager, *iscsi, iscsi.GetPath(), iscsi.mounter)
+	if err != nil {
+		glog.Errorf("iSCSIPersistentDisk: failed to setup")
+		return err
+	}
+	globalPDPath := iscsi.manager.MakeGlobalPDName(*iscsi)
+	// make mountpoint rw/ro work as expected
+	//FIXME revisit pkg/util/mount and ensure rw/ro is implemented as expected
+	mode := "rw"
+	if iscsi.readOnly {
+		mode = "ro"
+	}
+	iscsi.execCommand("mount", []string{"-o", "remount," + mode, globalPDPath, iscsi.GetPath()})
```

Comment:
I found passing  mount.FlagReadOnly to mount() still left the mountpoint rw. Before this mystery is resolved (see the FIXME text), I have to force a remount to ensure ro is indeed honored.